File: csharp/Chapter11/RunServer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Threading;
using com.db4o;

namespace com.db4o.dg2db4o.chapter11
{
    class RunServer : ServerConfiguration
    {
        public static void Main(string[] args)
        {
            RunServer s = new RunServer();
            Thread serverThread = new Thread(new ThreadStart(s.Run));
            serverThread.Priority = ThreadPriority.Highest;
            serverThread.Start();

            //AutoResetEvent autoEvent = new AutoResetEvent(false);
            //ThreadStarter threadStart = new ThreadStarter();
            //TimerCallback timerDelegate = new TimerCallback(threadStart.StartClients);
            //Timer delay = new Timer(timerDelegate, autoEvent, 5000, 0);
            //autoEvent.WaitOne(10000, false);
            //delay.Dispose();
            //Console.WriteLine("MAIN: Destroying timer, client threads should start...");

            RunClient c = new RunClient();
            RunReplicator r = new RunReplicator();
            Thread clientThread = new Thread(new ThreadStart(c.Run));
            Thread replicateThread = new Thread(new ThreadStart(r.Run));
            clientThread.Start();
            replicateThread.Start();
        }

        private bool stop = false;

        private void Run()
        {
            lock (this)
            {
                Console.WriteLine("Starting server...");
                File.Delete(REMOTEFILE);
                ObjectServer server = Db4o.OpenServer(REMOTEFILE, PORT);
                server.GrantAccess(USER, PASS);
                try
                {
                    while (!stop)
                    {
                        Monitor.Wait(this, 60000);
                        Console.WriteLine("SERVER: Server is listening ...");
                    }
                }
                catch (ThreadInterruptedException tie)
                {
                    Console.WriteLine("Thread Error!" + tie);
                }
                finally
                {
                    server.Close();
                }
            }
            Console.WriteLine("Server ends...");
        }

    }
}

File: java/src/com/db4o/dg2db4o/chapter11/RunServer.java

package com.db4o.dg2db4o.chapter11;

import java.io.*;
import com.db4o.*;

public class RunServer implements Runnable, ServerConfiguration {
    private boolean stop;
    
    // Can run as standalone process, or use Main.java to run as thread
    public static void main(String[] args) {
        RunServer runner = new RunServer();
        runner.run();
    }
    
    public void run() {
        synchronized (this) {
            new File(REMOTEFILE).delete();
            ObjectServer server = Db4o.openServer(REMOTEFILE, PORT);
            server.grantAccess(USER, PASS);
            try {
                while (!stop) {  // Out of band messages here later...
                    System.out.println("SERVER:[" + System.currentTimeMillis()
                    + "] Server's running... ");
                    this.wait(60000);
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                server.close();
            }
        }
    }
    
}
